File: src/sniff/report.py
import json
import re
from io import StringIO
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable

from rich.console import Console
from rich.text import Text

from sniff.models import SniffError

SOURCE_WIDTH = 76
INDENT = "\u3000\u3000"
PROVENANCE_LABELS = {
    "vale -> llm confirmed": "Vale → LLM",
    "ruff -> llm confirmed": "Ruff → LLM",
    "llm only": "LLM",
    "vale candidate": "Vale",
    "ruff candidate": "Ruff",
}
CONFIRMED_SOURCES = {
    "vale -> llm confirmed",
    "ruff -> llm confirmed",
    "llm only",
}
RULE_CODE = re.compile(r"^[A-Z]{3}[0-9]{3}$")


@dataclass(frozen=True)
class ReportFinding:
    path: str
    line: int
    column: int
    end_line: int
    end_column: int
    rule: str
    code: str
    severity: str
    source: str
    span: str
    message: str
    source_text: str | None = None


@dataclass(frozen=True)
class SourceRow:
    line: int | None
    text: str
    start_column: int = 0
    finding: bool = False
    blank: bool = False


def _integer(payload: dict[str, object], name: str, default: int | None = None) -> int:
    value = payload.get(name, default)
    if isinstance(value, bool) or not isinstance(value, int) or value < 1:
        raise SniffError(f"report finding has invalid {name!r}")
    return value


def _string(payload: dict[str, object], name: str) -> str:
    value = payload.get(name)
    if not isinstance(value, str) or not value:
        raise SniffError(f"report finding has invalid {name!r}")
    return value


def parse_report_findings(lines: Iterable[str]) -> list[ReportFinding]:
    findings: list[ReportFinding] = []
    for number, raw in enumerate(lines, 1):
        if not raw.strip():
            continue
        try:
            payload = json.loads(raw)
        except json.JSONDecodeError as exc:
            raise SniffError(f"invalid report JSON on input line {number}: {exc}") from exc
        if not isinstance(payload, dict):
            raise SniffError(f"report input line {number} must be a JSON object")
        line = _integer(payload, "line")
        column = _integer(payload, "column")
        span = _string(payload, "span")
        end_line = _integer(payload, "end_line", line)
        end_column = _integer(payload, "end_column", column + len(span))
        severity = _string(payload, "severity").lower()
        if severity not in {"suggestion", "warning", "error"}:
            raise SniffError(f"report finding has invalid severity {severity!r}")
        source = _string(payload, "source").lower()
        if source not in CONFIRMED_SOURCES:
            choices = ", ".join(sorted(CONFIRMED_SOURCES))
            raise SniffError(f"report finding has invalid source {source!r}; use: {choices}")
        source_text = payload.get("source_text")
        if source_text is not None and not isinstance(source_text, str):
            raise SniffError("report finding has invalid 'source_text'")
        code = _string(payload, "code")
        if not RULE_CODE.fullmatch(code):
            raise SniffError(f"report finding has invalid rule code {code!r}")
        findings.append(
            ReportFinding(
                path=_string(payload, "path"),
                line=line,
                column=column,
                end_line=end_line,
                end_column=end_column,
                rule=_string(payload, "rule"),
                code=code,
                severity=severity,
                source=source,
                span=span,
                message=_string(payload, "message"),
                source_text=source_text,
            )
        )
    return findings


def _source_lines(finding: ReportFinding, project_root: Path) -> list[str]:
    if finding.source_text is not None:
        return finding.source_text.splitlines()
    path = Path(finding.path).expanduser()
    if not path.is_absolute():
        path = project_root / path
    try:
        return path.read_text(encoding="utf-8", errors="replace").splitlines()
    except OSError:
        return [finding.span]


def _wrapped(text: str, width: int = SOURCE_WIDTH) -> list[tuple[int, str]]:
    if not text:
        return [(0, "")]
    rows: list[tuple[int, str]] = []
    start = 0
    while start < len(text):
        limit = min(start + width, len(text))
        end = limit
        if limit < len(text):
            split = text.rfind(" ", start + width // 2, limit + 1)
            if split > start:
                end = split
        rows.append((start, text[start:end].rstrip()))
        start = end
        while start < len(text) and text[start] == " ":
            start += 1
    return rows


def _clip_context(text: str, *, leading: bool) -> str:
    if len(text) <= SOURCE_WIDTH:
        return text
    if leading:
        return "… " + text[-(SOURCE_WIDTH - 2) :].lstrip()
    return text[: SOURCE_WIDTH - 2].rstrip() + " …"


def _target_rows(finding: ReportFinding, text: str) -> list[SourceRow]:
    wrapped = _wrapped(text)
    span_start = max(finding.column - 1, 0)
    span_end = max(span_start + len(finding.span), finding.end_column - 1)
    hits = [
        index
        for index, (offset, value) in enumerate(wrapped)
        if offset < span_end and offset + max(len(value), 1) > span_start
    ]
    if not hits:
        hits = [min(len(wrapped) - 1, span_start // SOURCE_WIDTH)]
    first, last = min(hits), max(hits)
    if last - first + 1 > 3:
        last = first + 2
    elif last == first and len(wrapped) > 1:
        if first + 1 < len(wrapped):
            last += 1
        elif first:
            first -= 1
    result: list[SourceRow] = []
    selected = wrapped[first : last + 1]
    for index, (offset, value) in enumerate(selected):
        rendered = value
        adjusted_offset = offset
        if index == 0 and first > 0:
            rendered = "… " + rendered
            adjusted_offset -= 2
        if index == len(selected) - 1 and last < len(wrapped) - 1:
            rendered = rendered.rstrip() + " …"
        result.append(
            SourceRow(
                line=finding.line if index == 0 else None,
                text=rendered,
                start_column=adjusted_offset,
                finding=True,
            )
        )
    return result


def _excerpt_rows(finding: ReportFinding, project_root: Path) -> list[SourceRow]:
    lines = _source_lines(finding, project_root)
    target_index = finding.line - 1
    target = lines[target_index] if 0 <= target_index < len(lines) else finding.span
    target_rows = _target_rows(finding, target)
    budget = max(0, 3 - len(target_rows))
    before: list[SourceRow] = []
    after: list[SourceRow] = []

    def context_row(index: int, *, leading: bool) -> SourceRow:
        value = lines[index]
        return SourceRow(
            line=index + 1,
            text=_clip_context(value, leading=leading),
            finding=finding.line <= index + 1 <= finding.end_line,
            blank=not bool(value),
        )

    if budget >= 2 and target_index > 0:
        before.append(context_row(target_index - 1, leading=True))
        if before[-1].blank and target_index > 1:
            before.insert(0, context_row(target_index - 2, leading=True))
        if any(not item.blank for item in before):
            budget -= 1
    index = target_index + 1
    while index < len(lines) and budget > 0:
        item = context_row(index, leading=False)
        after.append(item)
        if not item.blank:
            budget -= 1
        index += 1
    if budget > 0 and not before:
        index = target_index - 1
        while index >= 0 and budget > 0:
            item = context_row(index, leading=True)
            before.insert(0, item)
            if not item.blank:
                budget -= 1
            index -= 1
    return [*before, *target_rows, *after]


def _caret_row(finding: ReportFinding, row: SourceRow) -> str | None:
    if not row.finding:
        return None
    if row.line is not None and finding.end_line > finding.line:
        span_start = finding.column - 1 if row.line == finding.line else 0
        span_end = finding.end_column - 1 if row.line == finding.end_line else len(row.text)
    else:
        span_start = max(finding.column - 1, 0)
        span_end = max(span_start + len(finding.span), finding.end_column - 1)
    row_start = max(row.start_column, 0)
    visible_prefix = 2 if row.text.startswith("… ") else 0
    suffix = 2 if row.text.endswith(" …") else 0
    row_end = row_start + len(row.text) - visible_prefix - suffix
    overlap_start = max(span_start, row_start)
    overlap_end = min(span_end, row_end)
    if overlap_start >= overlap_end:
        return None
    padding = visible_prefix + overlap_start - row_start
    return " " * padding + "^" * max(1, overlap_end - overlap_start)


def _render_excerpt(finding: ReportFinding, project_root: Path) -> list[str]:
    rows = _excerpt_rows(finding, project_root)
    digits = len(str(max((row.line or 0 for row in rows), default=finding.line)))
    rendered: list[str] = []
    for row in rows:
        if row.line is None:
            marker, number = " ", " " * digits
        else:
            marker = "›" if row.line == finding.line else " "
            number = str(row.line).rjust(digits)
        rendered.append(f"    │ {marker} {number} │ {row.text}")
        caret = _caret_row(finding, row)
        if caret:
            rendered.append(f"    │   {' ' * digits} │ {caret}")
    return rendered


def _markdown_escape(value: str) -> str:
    return value.replace("\\", "\\\\").replace("[", "\\[").replace("]", "\\]")


def _location(
    path_value: str, line: int, project_root: Path, *, markdown: bool
) -> str:
    if path_value == "<stdin>":
        return f"<stdin>:{line}"
    path = Path(path_value).expanduser()
    absolute = path if path.is_absolute() else project_root / path
    absolute = absolute.resolve()
    try:
        label_path = absolute.relative_to(project_root.resolve()).as_posix()
    except ValueError:
        label_path = path_value
    label = _markdown_escape(f"{label_path}:{line}")
    if not markdown or not absolute.exists():
        return label
    return f"[{label}](<{absolute}:{line}>)"


def _italic(value: str) -> str:
    return value.replace("*", "\\*").replace("_", "\\_")


def _render_message(message: str, *, markdown: bool) -> list[str]:
    lines = [
        value
        for paragraph in message.split("\n")
        for _, value in _wrapped(paragraph)
    ]
    rendered: list[str] = []
    for index, line in enumerate(lines):
        prefix = f"{INDENT}└── " if index == 0 else f"{INDENT}    "
        styled = f"*{_italic(line)}*" if markdown else line
        rendered.append(prefix + styled)
    return rendered


def _terminal_report(
    findings: list[ReportFinding],
    project_root: Path,
    *,
    subject: str,
    color: bool,
) -> str:
    stream = StringIO()
    console = Console(
        file=stream,
        force_terminal=color,
        no_color=not color,
        color_system="standard" if color else None,
        width=120,
    )
    output = Text()
    output.append(" ___ _  _ ___ ___ ___\n", style="bold cyan")
    output.append("/ __| \\| |_ _| __| __|\n", style="bold cyan")
    output.append("\\__ \\ .` || || _|| _|\n", style="bold cyan")
    output.append("|___/_|\\_|___|_| |_|\n\n", style="bold cyan")
    subject_word = subject if len(findings) == 1 else f"{subject}s"
    file_count = len({finding.path for finding in findings})
    file_word = "file" if file_count == 1 else "files"
    output.append(
        f"{len(findings)} {subject_word} across {file_count} {file_word}\n",
        style="bold",
    )

    grouped: dict[tuple[str, int], list[ReportFinding]] = {}
    for finding in findings:
        grouped.setdefault((finding.path, finding.line), []).append(finding)
    for (path, line), items in grouped.items():
        items.sort(key=lambda item: (item.line, item.column))
        output.append("\n")
        output.append(
            _location(path, line, project_root, markdown=False),
            style="bold cyan underline",
        )
        output.append("\n")
        for index, finding in enumerate(items):
            if index:
                output.append("\n")
            marker = "└── " if finding.severity == "error" else INDENT
            provenance = PROVENANCE_LABELS[finding.source]
            severity_style = {
                "error": "bold red",
                "warning": "bold yellow",
                "suggestion": "bold cyan",
            }[finding.severity]
            output.append(marker)
            output.append(
                f"[{finding.severity.upper()} {finding.code}] {finding.rule}",
                style=severity_style,
            )
            output.append(" · ")
            output.append(f"via {provenance}\n\n", style="dim italic")
            for excerpt_line in _render_excerpt(finding, project_root):
                style = severity_style if "^" in excerpt_line else "dim"
                output.append(excerpt_line + "\n", style=style)
            output.append("\n")
            for message_line in _render_message(finding.message, markdown=False):
                output.append(message_line + "\n", style="italic")

    console.print(output, end="", soft_wrap=True)
    return stream.getvalue().rstrip("\n")


def render_report(
    findings: list[ReportFinding],
    project_root: Path,
    *,
    subject: str = "finding",
    markdown: bool = True,
    ansi: bool = False,
) -> str:
    if not findings:
        qualifier = "confirmed " if subject == "finding" else ""
        return f"sniff: no {qualifier}{subject}s"
    if not markdown:
        return _terminal_report(
            findings, project_root, subject=subject, color=ansi
        )
    grouped: dict[tuple[str, int], list[ReportFinding]] = {}
    for finding in findings:
        grouped.setdefault((finding.path, finding.line), []).append(finding)
    subject_word = subject if len(findings) == 1 else f"{subject}s"
    file_count = len({finding.path for finding in findings})
    file_word = "file" if file_count == 1 else "files"
    banner = [
        " ___ _  _ ___ ___ ___",
        "/ __| \\| |_ _| __| __|",
        "\\__ \\ .` || || _|| _|",
        "|___/_|\\_|___|_| |_|",
        "",
        f"{len(findings)} {subject_word} across {file_count} {file_word}",
    ]
    output = ["```text", *banner, "```"] if markdown else banner
    for (path, line), items in grouped.items():
        items.sort(key=lambda item: (item.line, item.column))
        location = _location(path, line, project_root, markdown=markdown)
        output.extend(["", location + ("  " if markdown else "")])
        for index, finding in enumerate(items):
            if index:
                output.append("")
            marker = "└── " if finding.severity == "error" else INDENT
            provenance = PROVENANCE_LABELS[finding.source]
            label = f"[{finding.severity.upper()} {finding.code}] {finding.rule}"
            if markdown:
                output.extend(
                    [
                        f"{marker}**{label}** · *via {provenance}*",
                        "",
                        "```text",
                        *_render_excerpt(finding, project_root),
                        "```",
                        "",
                        *_render_message(finding.message, markdown=True),
                    ]
                )
    return "\n".join(output)

File: rust/report.rs
use crate::{
    error::{Result, message},
    models::{ReportFinding, Severity},
};
use regex::Regex;
use std::{
    collections::{BTreeMap, BTreeSet},
    fs,
    path::{Path, PathBuf},
};
const WIDTH: usize = 76;
const INDENT: &str = "　　";
fn provenance(s: &str) -> Option<&'static str> {
    match s {
        "vale -> llm confirmed" => Some("Vale → LLM"),
        "ruff -> llm confirmed" => Some("Ruff → LLM"),
        "llm only" => Some("LLM"),
        "vale candidate" => Some("Vale"),
        "ruff candidate" => Some("Ruff"),
        _ => None,
    }
}
pub fn parse(input: &str, confirmed: bool) -> Result<Vec<ReportFinding>> {
    let mut out = vec![];
    let rule_code = Regex::new(r"^[A-Z]{3}[0-9]{3}$").unwrap();
    for (number, line) in input.lines().enumerate() {
        if line.trim().is_empty() {
            continue;
        }
        let f: ReportFinding = serde_json::from_str(line).map_err(|e| {
            message(format!(
                "invalid report JSON on input line {}: {e}",
                number + 1
            ))
        })?;
        if f.line < 1
            || f.column < 1
            || f.path.is_empty()
            || f.rule.is_empty()
            || f.span.is_empty()
            || f.message.is_empty()
        {
            return Err(message("report finding has invalid required field"));
        }
        if !rule_code.is_match(&f.code) {
            return Err(message(format!(
                "report finding has invalid rule code {:?}",
                f.code
            )));
        }
        if provenance(&f.source).is_none()
            || confirmed && matches!(f.source.as_str(), "vale candidate" | "ruff candidate")
        {
            return Err(message(format!(
                "report finding has invalid source {:?}",
                f.source
            )));
        }
        out.push(f.normalize())
    }
    Ok(out)
}
fn lines(f: &ReportFinding, root: &Path) -> Vec<String> {
    if let Some(x) = &f.source_text {
        return x.lines().map(str::to_owned).collect();
    }
    let p = PathBuf::from(&f.path);
    let p = if p.is_absolute() { p } else { root.join(p) };
    fs::read_to_string(p)
        .map(|x| x.lines().map(str::to_owned).collect())
        .unwrap_or_else(|_| vec![f.span.clone()])
}
fn wrap(text: &str) -> Vec<(usize, String)> {
    if text.is_empty() {
        return vec![(0, String::new())];
    }
    let chars: Vec<char> = text.chars().collect();
    let mut out = vec![];
    let mut start = 0;
    while start < chars.len() {
        let limit = (start + WIDTH).min(chars.len());
        let mut end = limit;
        if limit < chars.len()
            && let Some(i) = chars[start + WIDTH / 2..=limit]
                .iter()
                .rposition(|c| *c == ' ')
        {
            end = start + WIDTH / 2 + i
        }
        out.push((
            start,
            chars[start..end]
                .iter()
                .collect::<String>()
                .trim_end()
                .into(),
        ));
        start = end;
        while start < chars.len() && chars[start] == ' ' {
            start += 1
        }
    }
    out
}
fn excerpt(f: &ReportFinding, root: &Path) -> Vec<String> {
    let all = lines(f, root);
    let idx = f.line - 1;
    let target = all.get(idx).cloned().unwrap_or_else(|| f.span.clone());
    let wrapped = wrap(&target);
    let start = f.column - 1;
    let end = (f.end_column.unwrap() - 1).max(start + f.span.chars().count());
    let mut hits: Vec<_> = wrapped
        .iter()
        .enumerate()
        .filter(|(_, (o, s))| *o < end && *o + s.chars().count().max(1) > start)
        .map(|(i, _)| i)
        .collect();
    if hits.is_empty() {
        hits.push((start / WIDTH).min(wrapped.len() - 1))
    }
    let first = *hits.first().unwrap();
    let last = (*hits.last().unwrap()).min(first + 2);
    let mut rows: Vec<(Option<usize>, String, usize, bool)> = vec![];
    let budget = 3 - (last - first + 1);
    if budget >= 2 && idx > 0 {
        rows.push((Some(idx), all[idx - 1].clone(), 0, false));
    }
    for (n, (offset, text)) in wrapped[first..=last].iter().enumerate() {
        let mut shown = text.clone();
        let mut adjusted = *offset;
        if n == 0 && first > 0 {
            shown = format!("… {shown}");
            adjusted = adjusted.saturating_sub(2)
        }
        if n == last - first && last + 1 < wrapped.len() {
            shown = format!("{} …", shown.trim_end())
        }
        rows.push((
            if n == 0 { Some(f.line) } else { None },
            shown,
            adjusted,
            true,
        ));
    }
    let used = rows
        .iter()
        .filter(|x| x.0.is_some() && !x.1.is_empty())
        .count();
    let mut next = idx + 1;
    while next < all.len() && rows.len().saturating_sub(used) + used < 3 {
        rows.push((
            Some(next + 1),
            all[next].chars().take(WIDTH).collect(),
            0,
            next < f.end_line.unwrap(),
        ));
        next += 1
    }
    let digits = rows
        .iter()
        .filter_map(|x| x.0)
        .max()
        .unwrap_or(f.line)
        .to_string()
        .len();
    let mut out = vec![];
    for (line, text, offset, finding) in rows {
        let marker = if line == Some(f.line) { "›" } else { " " };
        let number = line
            .map(|x| format!("{x:>digits$}"))
            .unwrap_or_else(|| " ".repeat(digits));
        out.push(format!("    │ {marker} {number} │ {text}"));
        if finding {
            let row_start = offset;
            let row_end = row_start + text.chars().count();
            let a = start.max(row_start);
            let b = end.min(row_end);
            if a < b {
                out.push(format!(
                    "    │   {} │ {}{}",
                    " ".repeat(digits),
                    " ".repeat(a - row_start),
                    "^".repeat(b - a)
                ));
            }
        }
    }
    out
}
fn location(f: &ReportFinding, root: &Path, markdown: bool) -> String {
    if f.path == "<stdin>" {
        return format!("<stdin>:{}", f.line);
    }
    let p = PathBuf::from(&f.path);
    let abs = if p.is_absolute() { p } else { root.join(p) }
        .canonicalize()
        .unwrap_or_else(|_| root.join(&f.path));
    let label = abs
        .strip_prefix(root)
        .map(|x| x.display().to_string())
        .unwrap_or_else(|_| f.path.clone());
    if markdown && abs.exists() {
        format!("[{}:{}](<{}:{}>)", label, f.line, abs.display(), f.line)
    } else {
        format!("{}:{}", label, f.line)
    }
}
fn rendered_message(message: &str, markdown: bool) -> String {
    message
        .split('\n')
        .flat_map(|paragraph| wrap(paragraph).into_iter().map(|(_, line)| line))
        .enumerate()
        .map(|(index, line)| {
            let prefix = if index == 0 {
                format!("{INDENT}└── ")
            } else {
                format!("{INDENT}    ")
            };
            if markdown {
                format!("{prefix}*{}*", line.replace('*', "\\*").replace('_', "\\_"))
            } else {
                format!("{prefix}{line}")
            }
        })
        .collect::<Vec<_>>()
        .join("\n")
}
pub fn render(
    findings: &[ReportFinding],
    root: &Path,
    subject: &str,
    markdown: bool,
    color: bool,
) -> String {
    if findings.is_empty() {
        return format!(
            "sniff: no {}{}s",
            if subject == "finding" {
                "confirmed "
            } else {
                ""
            },
            subject
        );
    }
    let files = findings
        .iter()
        .map(|x| &x.path)
        .collect::<BTreeSet<_>>()
        .len();
    let mut out = if markdown {
        format!(
            "```text\n ___ _  _ ___ ___ ___\n/ __| \\| |_ _| __| __|\n\\__ \\ .` || || _|| _|\n|___/_|\\_|___|_| |_|\n\n{} {}{} across {} file{}\n```",
            findings.len(),
            subject,
            if findings.len() == 1 { "" } else { "s" },
            files,
            if files == 1 { "" } else { "s" }
        )
    } else {
        format!(
            " ___ _  _ ___ ___ ___\n/ __| \\| |_ _| __| __|\n\\__ \\ .` || || _|| _|\n|___/_|\\_|___|_| |_|\n\n{} {}{} across {} file{}",
            findings.len(),
            subject,
            if findings.len() == 1 { "" } else { "s" },
            files,
            if files == 1 { "" } else { "s" }
        )
    };
    let mut groups: BTreeMap<(&str, usize), Vec<&ReportFinding>> = BTreeMap::new();
    for f in findings {
        groups.entry((&f.path, f.line)).or_default().push(f)
    }
    for (_, mut items) in groups {
        items.sort_by_key(|x| (x.line, x.column));
        out.push_str("\n\n");
        out.push_str(&location(items[0], root, markdown));
        if markdown {
            out.push_str("  ")
        }
        for f in items {
            let label = format!(
                "[{} {}] {}",
                f.severity.as_str().to_uppercase(),
                f.code,
                f.rule
            );
            let marker = if f.severity == Severity::Error {
                "└── "
            } else {
                INDENT
            };
            if markdown {
                out.push_str(&format!(
                    "\n{marker}**{label}** · *via {}*\n\n```text\n{}\n```\n\n{}",
                    provenance(&f.source).unwrap(),
                    excerpt(f, root).join("\n"),
                    rendered_message(&f.message, true)
                ));
            } else {
                let styled = if color {
                    format!("\x1b[1m{label}\x1b[0m")
                } else {
                    label
                };
                out.push_str(&format!(
                    "\n{marker}{styled} · via {}\n\n{}\n\n{}",
                    provenance(&f.source).unwrap(),
                    excerpt(f, root).join("\n"),
                    rendered_message(&f.message, false)
                ));
            }
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn payload(source: &str) -> String {
        format!(
            r#"{{"path":"prompt.md","line":1,"column":1,"rule":"test-rule","code":"TST001","severity":"warning","source":"{source}","span":"Must","message":"Unbounded."}}"#
        )
    }

    #[test]
    fn rejects_bad_json_code_severity_and_unconfirmed_source() {
        assert!(
            parse("not json", true)
                .unwrap_err()
                .to_string()
                .contains("input line 1")
        );
        assert!(
            parse(&payload("vale candidate"), true)
                .unwrap_err()
                .to_string()
                .contains("invalid source")
        );
        let bad_code = payload("llm only").replace("TST001", "bad");
        assert!(
            parse(&bad_code, true)
                .unwrap_err()
                .to_string()
                .contains("rule code")
        );
        let bad_severity = payload("llm only").replace("warning", "fatal");
        assert!(parse(&bad_severity, true).is_err());
    }

    #[test]
    fn defaults_end_positions_and_renders_both_formats() {
        let findings = parse(&payload("llm only"), true).unwrap();
        assert_eq!(findings[0].end_line, Some(1));
        assert_eq!(findings[0].end_column, Some(5));
        let terminal = render(&findings, Path::new("."), "finding", false, false);
        assert!(terminal.contains("[WARNING TST001] test-rule · via LLM"));
        assert!(terminal.contains("^^^^"));
        let markdown = render(&findings, Path::new("."), "finding", true, false);
        assert!(markdown.contains("**[WARNING TST001] test-rule**"));
        assert!(markdown.contains("```text"));
    }

    #[test]
    fn empty_reports_distinguish_candidates_and_findings() {
        assert_eq!(
            render(&[], Path::new("."), "finding", false, false),
            "sniff: no confirmed findings"
        );
        assert_eq!(
            render(&[], Path::new("."), "candidate", false, false),
            "sniff: no candidates"
        );
    }

    #[test]
    fn long_diagnoses_use_a_hanging_indent() {
        let message = "The allowed degree of editing is not defined, so different tools can make materially different changes while claiming to clean the reports lightly.";
        let rendered = rendered_message(message, false);
        let lines: Vec<_> = rendered.lines().collect();
        assert!(lines.len() > 1);
        assert!(lines[0].starts_with("　　└── "));
        assert!(lines[1].starts_with("　　    "));
    }
}
